### DISCUSSION TITLE
struct having name like `MyData_T_int`

### DISCUSSION THREAD

**Discussion by impopular-guy** [vlang/v#17239] [General]
I purposefully named the struct `MyData_T_int` to see what happens. This fails to run but C code is generated without errors. It seems like the definition of `MyData[int]` is overwritten by `MyData_T_int`, so there ends up only one struct definition which causes the error.

```v
struct MyData[T]{
    a T
}

struct MyData_T_int{
    b string
}

fn main() {
    d1 := MyData[int]{12}
    println(d1)
    d2 := MyData_T_int{'qwerty'}
    println(d2)
}
```

The code runs if fields of both struct have same name and same type or type is castable(like int to f32).

How should such a problem be solved? I tried searching the internet but didn't find anything. I am not very knowledgeable about compilers and not sure how other languages solve such an issue?

Possible soln that come to my mind 1) not allow such name 2) add some suffix like `MyData_T_int_suffix` 3) ignore?

Btw, great language, keep up the great work.

**Comment by JalonSolov** [comment]
Interesting way of hacking the compiler.  :-)

1) Not sure how the name could be prevented.  V wouldn't know there was a problem until backend code generation time.

2) This might work, but we would have to study the effects.

3) No, I don't think it should be ignored, since it can obviously cause problems with name collisions.

**Comment by i582** [comment]
Thanks for the interesting find!
I think all generated names should be prefixed with `__` which should be forbid for user code.

**Reply by JalonSolov** [reply]
... except some C compilers treat variables with `__` prefix differently than regular variables.

**Reply by i582** [reply]
> ... except some C compilers treat variables with `__` prefix differently than regular variables.

ok, suffix

**Comment by medvednikov** [comment]
In the original V, _ was banned in struct names for this reason. Also nice for the one way rule.

I wish I kept this rule.

**Reply by Delta456** [reply]
I mean we can simply change the C code generation for this. There is no need to ban using `_`.

**Comment by spytheman** [comment]
How about just banning `_T_` in struct names?

Edit: it would need to be a ban for `_CAPITAL_`, not just `_T_` .

**Reply by Delta456** [reply]
> How about just banning `_T_` in struct names?
> 
> Edit: it would need to be a ban for `_CAPITAL_`, not just `_T_` .

I feel this is quite restricting and the user should be able to name what they want too.

**Comment by Wajinn** [comment]
It's better to know about such hacks, even if inconvenient, than to not know.  This way they can be dealt with in some way.
